src/s3-copy.py

```diff
@@ -26,17 +26,63 @@
 cycling_interval = config_dict.get('cycling_interval')
 # read source info
 sources = config_dict.get('source')
+# read destination info
+destination = config_dict.get('destination')
+
+#----- create source clients
+print('Opening aws resources')
+destination_bucket = boto3.Session(profile_name=destination.get('credential_profile')).resource('s3').Bucket(destination["bucket"])
+destination_fn_template = destination.get('path')+destination["file_template"]["prefix"]+destination["file_template"]["root"]+destination["file_template"]["suffix"]
+print('...Done')
+
+#----- open log files
+print('Opening loggers')
+dtg = datetime.now().strftime('%Y%m%d%H%M%S')
+logger_success = open(config_dict["logging"]["success"].format(dtg=dtg),'wt')
+logger_missing = []
+si = 0
+for source in sources.values():
+  sn = source["name"].replace(' ','_')
+  logger_missing.append( open(config_dict["logging"]["missing"].format(dtg=dtg,source_num=si,source_name=sn), 'wt') )
+  si = si + 1
+print('...Done')
 
 #----- cycle through the date range
+print('Top of the time loop')
 while not(dr.at_end()):
+  #----- create destination path
   current_cycle = dr.current
+  destination_key = current_cycle.strftime(destination_fn_template)
   print(current_cycle)
-  dr.increment(seconds = cycling_interval)
-  #----- cycle through the sources
-  for source in sources.items():
-    source_name = source[1].get('name')
-    print(f"using {source_name} at time {current_cycle}")
 
-print('at the end of the time loop')
+  #----- cycle through the sources
+  si = -1
+  for source in sources.values():
+    si = si + 1
+    # construct source filename
+    fn_template = source["file_template"]["prefix"]+source["file_template"]["root"]+source["file_template"]["suffix"]
+    source_dict = {"Bucket" : source["bucket"], "Key" : current_cycle.strftime(source["key"]+fn_template)}
+    # if source file exists, then copy
+    try:
+      destination_bucket.copy( source_dict, destination_key)
+    except botocore.exceptions.ClientError as e:
+      if e.response['Error']['Code'] == "404":
+        # log that the file is missing and try next source
+        logger_missing[si].write("s3://{}/{}\n".format(source_dict["Bucket"], source_dict["Key"]));
+        continue
+      else:
+        # something else went wrong
+        raise
+    else:
+      # log success message
+      logger_success.write("s3://{}/{} -> s3://{}/{}\n".format(source_dict["Bucket"], source_dict["Key"], destination["bucket"], destination_key))
```

Comment:
I would create a single function for this and use Python list comprehensions to do the looping. This kind of looping within Python is sub-optimal.

Reply:
agreed. My plan was to tackle it when the package grows further.